lib/libvfio-user.c

```diff
@@ -459,6 +461,182 @@ handle_device_get_region_info(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+    elem->size = size;
+    elem->flags = flags;
+    elem->datamatch = datamatch;
+    LIST_INSERT_HEAD(&vfu_reg->subregions, elem, entry);
+
+    return 0;
+}
+
+static void
+free_regions(vfu_ctx_t *vfu_ctx)
+{
+    size_t index = 0;
+
+    assert(vfu_ctx != NULL);
+
+    for (index = 0; index < VFU_PCI_DEV_NUM_REGIONS; index++) {
+        vfu_reg_info_t *vfu_reg = &vfu_ctx->reg_info[index];
+
+        while (!LIST_EMPTY(&vfu_reg->subregions)) {
+            ioeventfd_t *n = LIST_FIRST(&vfu_reg->subregions);
+            LIST_REMOVE(n, entry);
+            free(n);
+        }
+    }
+    free(vfu_ctx->reg_info);
+}
+
+/*
+ * This function is used to add fd's to the fd return array and gives you back
+ * the index of the fd that has been added. If the fd is already present it will
+ * return the index to that duplicate fd to reduce the number of fd's sent.
+ */
+static int
+add_fd_index(int *out_fds, size_t *nr_out_fds, int fd_search)
+{
+    size_t i = 0;
+
+    assert(out_fds != NULL);
+    assert(nr_out_fds != NULL);
+
+    for (i = 0; i < *nr_out_fds; i++) {
+        if (out_fds[i] == fd_search) {
+            return i;
+        }
+    }
+
+    out_fds[*nr_out_fds] = fd_search;
+    (*nr_out_fds)++;
+
+    return *nr_out_fds - 1;
+}
+
+static int
+handle_device_get_region_io_fds(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+{
+    size_t max_sent_sub_regions = 0;
+    vfu_reg_info_t *vfu_reg = NULL;
+    vfio_user_region_io_fds_reply_t *reply = NULL;
+    vfio_user_sub_region_ioeventfd_t *ioefd = NULL;
+    vfio_user_region_io_fds_request_t *req = NULL;
+    ioeventfd_t *sub_reg = NULL;
+    size_t nr_sub_reg = 0;
+    size_t i = 0;
+
+    assert(vfu_ctx != NULL);
+    assert(msg != NULL);
+    assert(msg->out_fds == NULL);
+
+    if (msg->in_size != sizeof(vfio_user_region_io_fds_request_t)) {
+        return ERROR_INT(EINVAL);
+    }
+
+    req = msg->in_data;
+
+    if (req->flags != 0 || req->count != 0) {
+        return ERROR_INT(EINVAL);
+    }
+
+    if (req->index >= vfu_ctx->nr_regions) {
+        vfu_log(vfu_ctx, LOG_DEBUG, "bad region index %d in get region io fds "
+                "info", req->index);
+        return ERROR_INT(EINVAL);
+    }
+
+    if (req->index >= VFU_PCI_DEV_NUM_REGIONS) {
+        return ERROR_INT(EINVAL);
+    }
+
+    vfu_reg = &vfu_ctx->reg_info[req->index];
+
+    if (vfu_reg->fd == -1) {
+        return ERROR_INT(EINVAL);
+    }
+
+    LIST_FOREACH(sub_reg, &vfu_reg->subregions, entry) {
+        nr_sub_reg++;
+    }
+
+    if (req->argsz < sizeof(vfio_user_region_io_fds_reply_t) ||
+        req->argsz > SERVER_MAX_DATA_XFER_SIZE) {
+        return ERROR_INT(EINVAL);
+    }
+
+    max_sent_sub_regions = MIN((req->argsz -
+                                sizeof(vfio_user_region_io_fds_reply_t)) /
+                                sizeof(vfio_user_sub_region_ioeventfd_t),
+                                nr_sub_reg);
+    max_sent_sub_regions = (max_sent_sub_regions >= nr_sub_reg) ? nr_sub_reg : 0;
```

Comment:
this is a little weird now. Really you're saying "if we can't fit nr_sub_reg, then don't add anything". But what you do now looks a bit confusing, and in fact you end up doing a calloc of size 0 in this case, on line :617

Instead, take a look at how we do this in `handle_device_get_region_info`, and do something like this:

```
region_array_size = nr_sub_reg * sizeof(vfio_user_sub_region_ioeventfd_t);
...
msg->out_size = MIN(sizeof(*reply) + region_array_size, in_info->argsz);
msg->out_data = calloc(1, msg->out_size);

...

reply = msg->out_data;
reply->argsz = sizeof(*reply) + region_array_size;
...
/* Only provide sub-region info if it fits. */
if (in_info->argsz >= out_info->argsz) {
   /* alloc/set ->out_fds here too, then output each region... */
   sub_reg = LIST_FIRST(...
}
```

make sense?

Reply:
yep fixing.